### PR TITLE
Fix "The stock counter is not updated when actions on cart products are performed "

### LINKED ISSUE #18289: BO - New order page - The stock counter is not updated until a page refresh when a product is added to order
**Describe the bug**

In BO => Orders => Orders => Add an Order when a product is added the quantity is not updated until we refresh the page

#### Expected behavior

The quantity should be updated

#### Steps to Reproduce

1. Go to BO => Orders => Add an Orders
2. Select a customer
3. Select a product and click on "add to cart"
4. See error the quantity is not updated

**Screenshots**

capture d'écran_8

#### Additional information

* PrestaShop version: 1.7.7.x, 1.7.6.x
* PHP version: 7.2

### PR DESCRIPTION
**Pull request by arouiadib**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | (In Orders/ Create new order page ) This PR updates the stock counter of a product/combination when an action is performed on cart for that product/combination. An action can be: Adding to cart, updating quantity in cart, or deleting from cart. 
| Type?         | bug fix
| Category?     | BO 
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes #18289
| How to test?  | In migrated Orders/ Create new order page, try to perform an action on a product/combination among the above mentioned actions, you will see that the stock counter is updated. 

Hi @marionf , @NeOMakinG ,
I hope you are doing well. 
I have created a new PR because I didn't manage to rebase on develop without spoiling my branch. I will try be more careful in the future. 
Thanks again for your time reviewing this for 2nd time.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/18510)

### CODE REVIEW

File: admin-dev/themes/new-theme/js/pages/order/create/cart-editor.js [hunk 1]
```diff
@@ -140,7 +140,7 @@ export default class CartEditor {
       productId: product.productId,
       attributeId: product.attributeId,
       customizationId: product.customizationId,
-    }).then((cartInfo) => EventEmitter.emit(eventMap.productRemovedFromCart, cartInfo))
+    }).then((cartInfo) => EventEmitter.emit(eventMap.productRemovedFromCart, {'cartInfo': cartInfo , 'product' : product }))
```
Comment (Progi1984) [inline admin-dev/themes/new-theme/js/pages/order/create/cart-editor.js]:
```suggestion
    }).then((cartInfo) => EventEmitter.emit(eventMap.productRemovedFromCart, {cartInfo, product}))
```

File: admin-dev/themes/new-theme/js/pages/order/create/cart-editor.js [hunk 2]
```diff
@@ -177,7 +177,7 @@ export default class CartEditor {
       newQty: product.newQty,
       attributeId: product.attributeId,
       customizationId: product.customizationId,
-    }).then((cartInfo) => EventEmitter.emit(eventMap.productQtyChanged, cartInfo))
+    }).then((cartInfo) => EventEmitter.emit(eventMap.productQtyChanged, {'cartInfo': cartInfo , 'product' : product }))
```
Comment (Progi1984) [inline admin-dev/themes/new-theme/js/pages/order/create/cart-editor.js]:
```suggestion
    }).then((cartInfo) => EventEmitter.emit(eventMap.productQtyChanged, {cartInfo, product}))
```

File: admin-dev/themes/new-theme/js/pages/order/create/product-manager.js [hunk 3]
```diff
@@ -328,4 +331,102 @@ export default class ProductManager {
...
     };
   }
-}
+
+  /**
+   * Updates the stock when the product is added to cart in "create new order" page
+   *
+   * @private
+   */
+  _updateStockOnProductAdd() {
```
Comment (Progi1984) [inline admin-dev/themes/new-theme/js/pages/order/create/product-manager.js]:
```suggestion
  updateStockOnProductAdd() {
```

File: admin-dev/themes/new-theme/js/pages/order/create/product-manager.js [hunk 4]
```diff
@@ -328,4 +331,102 @@ export default class ProductManager {
...
+      }
+    }
+  }
+
+  /**
+   * Updates the stock when the product is removed from cart in Orders/"create new order page"
+   *
+   * @private
+   */
+  _updateStockOnProductRemove(product){
```
Comment (Progi1984) [inline admin-dev/themes/new-theme/js/pages/order/create/product-manager.js]:
```suggestion
  updateStockOnProductRemove(product){
```

File: admin-dev/themes/new-theme/js/pages/order/create/product-manager.js [hunk 5]
```diff
@@ -328,4 +331,102 @@ export default class ProductManager {
...
+      }
+    }
+  }
+
+  /**
+   * Updates the stock when the quantity of product is changed from cart in Orders/"create new order page"
+   *
+   * @private
+   */
+  _updateStockOnQtyChange(product){
```
Comment (Progi1984) [inline admin-dev/themes/new-theme/js/pages/order/create/product-manager.js]:
```suggestion
  updateStockOnQtyChange(product){
```

File: admin-dev/themes/new-theme/js/pages/order/create/product-manager.js [hunk 6]
```diff
@@ -328,4 +331,102 @@ export default class ProductManager {
...
+      }
+    }
+  }
+
+  /**
+   * Renders stock depending on product or combination
+   *
+   * @private
+   */
+  _renderStockCounter(productId, combinationId, key){
```
Comment (Progi1984) [inline admin-dev/themes/new-theme/js/pages/order/create/product-manager.js]:
```suggestion
  renderStockCounter(productId, combinationId, key){
```

**Review by Progi1984 — COMMENTED** [review]
Some feedbacks

**Comment by NeOMakinG** [comment]
Also there are some ESLint errors to be fixed :)

**Review by PierreRambaud — DISMISSED** [review]
Eslint need to be fixed

**Comment by arouiadib** [comment]
Thank you guys.
Eslint is really a good man :D, it helps you create clean code while learning new things. Thanks for the guy who integrated it in Prestashop tech stack .
just few things please:
- Should I remove the comments in the PR? they are too many.
- I know that I should add test, I am working on it. What if I do them in separate PR :sweat_smile: ?
Again thanks for your time guys :heart:

**Comment by NeOMakinG** [comment]
Hey @arouiadib, thanks to @PierreRambaud for ESLint merged like 2 or 3 months ago, also recently, stylelint has been merged for the SCSS part :)

I think comments are fine to be honest, but if you really care about readability, you can do as I do : adding a new line before every comments, aerate the code !

I can't say anything about tests writing, what do you think @PrestaShop/prestashop-qa , @boubkerbribri ?

I think if you want to write test, that should be in the same PR as they would require to be merged together. We recently faced that situation with @sowbiba and @boubkerbribri that's why we should wait their advices.

Thanks for you contribution

**Comment by boubkerbribri** [comment]
> I can't say anything about tests writing, what do you think @PrestaShop/prestashop-qa , @boubkerbribri ?

I think that @arouiadib is talking about Unit tests. If so it would be better to have it in this PR, maybe @PrestaShop/prestashop-core-developers can better talk about this.

For tests UI (with puppeteer), @arouiadib  if you want to add some tests, you can do it in a new PR.

> I think if you want to write test, that should be in the same PR as they would require to be merged together. We recently faced that situation with @sowbiba and @boubkerbribri that's why we should wait their advices.

It was different, @sowbiba PR was changing some visual elements, and tests had to be adapted.
I don't think this PR make UI tests fail.

**Comment by Progi1984** [comment]
@boubkerbribri Not agree, if tests are relative to this PR, they must be in the same PR and not in a dedicated PR.

**Comment by SimonGrn** [comment]
Our puppeteer tests are not impacted by this PR. We are not concerned here, I'm pretty sure @arouiadib was talking about unit tests.

**Comment by PierreRambaud** [comment]
> @boubkerbribri Not agree, if tests are relative to this PR, they must be in the same PR and not in a dedicated PR.

Agree, UI tests must be done in the same pull request! 
- If we revert this one, we will have to find all pr with tests related to it
- it's a very bad practice
- it's hard to understand why tests are changing

**Comment by boubkerbribri** [comment]
> @boubkerbribri Not agree, if tests are relative to this PR, they must be in the same PR and not in a dedicated PR.

Still talking about puppeteer tests : 

- For new tests :  of course it would be better if it is in this PR, but it's not an obligation. it could be done also after validating the PR manually (assuming that the PR does not change how the UI work and the test will work even if the PR is reverted).

- if a fix on an existing test had to be done due to this PR, the fix must be done in this PR (and like @SimonGrn said, this PR has no impact on tests already done)

**Comment by arouiadib** [comment]
I thank you very much guys for the informative discussion. :heart:
I will be back soon to ask you reviewing my first front end tests ever. 
:) with best regards

**Comment by arouiadib** [comment]
Guys, I think I should leave this branch in peace and commit my first tests in other PR. 
Thank you for understanding.

**Comment by faisal616** [comment]
‫أُرسلت من الـ iPhone‬

**Comment by Junebyun** [comment]
Hi guys! Is there any wording to check?

**Review by NeOMakinG — APPROVED** [review]

**Comment by arouiadib** [comment]
Hi @Junebyun ,
Thank you very much for your time. 
No, there is no wording for this PR.

**Review by Progi1984 — APPROVED** [review]

**Comment by SD1982** [comment]
Hi @arouiadib It seems that @PierreRambaud has requested some change can you check this please ?

**Comment by PierreRambaud** [comment]
> Hi @arouiadib It seems that @PierreRambaud has requested some change can you check this please ?

Changes were applied, I dismiss my review ;)

**Comment by SD1982** [comment]
LGTM Thanks @arouiadib !

**Comment by Progi1984** [comment]
Thanks @arouiadib & @SD1982

**Pull request merged**